src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -165,6 +171,52 @@ func (s *serviceState) Health() (*rpc.NodeHealthResult_, bool) {
 	return v, v != nil
 }
 
+func (s *serviceState) DBForWriteRPCWithLimit() (
+	db storage.Database, dbInitialized bool, rpcDoesNotExceedLimit bool) {
+	s.Lock()
+	defer s.Unlock()
```

Comment:
Just FYI, we really should use an RLock and an atomic for the tracking of RPC exceeding limit, this an exclusive lock which long term might not be great for core write path.

Reply:
@robskillington Yeah I agree an atomic would probably be better, but the exclusive lock is only ever taken if you explicitly opt into this feature so it will never impact users who don't choose to use it.

Also I just didn't want to spend the time futzing around with the atomics right now when its really not a bottleneck (even if the node is doing a million writes/s we're talking about 7-8k locks which is really not that much).